### PR TITLE
resolve doc todos - oliver

### PR DESCRIPTION
**Pull request by kw-fn**
bumped into a couple `TODO`s in the comments while
studying the code, hopefully overall improvement here

:cake:

### CODE REVIEW

**Comment by RustyYato** [comment]
Thanks for the PR! I'll take a closer look at this tommorow, looks good overall

File: core/src/lib.rs [hunk 1]
```diff
@@ -131,67 +130,67 @@ unsafe impl<T> Field for Identity<T> {
...
 
-/// Projects a type to the given field
+/// Projects a type to a given `Field`
 pub trait ProjectTo<F: Field> {
-    /// The projection of the type, can be used to directly access the field
+    /// - directly access to the generated `Field`
     type Projection;
 
-    /// projects to the given field
+    /// - project to a given `Field`
```
Comment (RustyYato) [inline core/src/lib.rs:138]:
```suggestion
    /// project to a given `Field`
```

File: core/src/lib.rs [hunk 2]
```diff
@@ -131,67 +130,67 @@ unsafe impl<T> Field for Identity<T> {
...
+/// * For `Option<T>`, if it is `Some`, the safety condition on `T` applies
 ///
-/// * For `*const T`, `*mut T`, and `NonNull<T>`, it's the same as
-///   `project_raw`/`project_raw_mut`
-/// * For `Option<T>`, if it is `Some`, then the safety condtion on `T`
-///   applies, otherwise there is no safety condition
+/// * No safety condition otherwise applies
 pub trait UncheckedProjectTo<F: Field> {
-    /// The projection of the type, can be used to directly access the field
+    /// - direct access to a `Copy`/`Clone` of the generated `Field`
```
[the remark this answers is not in the document]
Comment (RustyYato) [reply core/src/lib.rs:154]:
```suggestion
    /// direct access to via a pointer-like type to the field
```

File: core/src/lib.rs [hunk 3]
```diff
@@ -131,67 +130,67 @@ unsafe impl<T> Field for Identity<T> {
...
 
-/// Projects a type to the given field list
+/// Projects a type to the given `Field` list
 pub trait ProjectAll<Parent, F> {
-    /// The projection of the type, can be used to directly access the field
+    /// - direct access to the generated `Field`
     type Projection;
 
-    /// projects to the given field list
+    /// - projection to the given `Field` list
```
Comment (RustyYato) [inline core/src/lib.rs:187]:
```suggestion
    /// projection to the given `Field` list
```

File: core/src/lib.rs [hunk 4]
```diff
@@ -131,67 +130,67 @@ unsafe impl<T> Field for Identity<T> {
...
     fn project_all(self, field_list: F) -> Self::Projection;
 }
 
-/// Represents a field of some `Parent` type.
-///
-/// You can derive this trait for all fields of `Parent` by using:
+/// A generated representation of some `Parent` types `struct`. `Parent`
+/// represents the type where the field came from , `Type` represents the type
```
Comment (RustyYato) [inline core/src/lib.rs:192]:
```suggestion
/// A generated representation of some `Parent` types `struct` or `union`. `Parent`
/// represents the type where the field came from, `Type` represents the type
```

File: core/src/lib.rs [hunk 5]
```diff
@@ -208,14 +207,11 @@ pub trait ProjectAll<Parent, F> {
...
-/// * `Type` must represent the type of the field itself
-/// * `project_raw` and `project_raw_mut` must only access the given field
-/// * `name` must return an iterator that yields all of the fields from `Parent`
-///   to the given field,
-/// * You must not override `field_offset`, `inverse_project_raw`,
-///   or `inverse_project_raw_mut`
-/// ex.
+/// * `project_raw` and `project_raw_mut` must only access the given `Field`
+/// * `name` must return an iterator yielding all the fields from `Parent` to
+///   the given field
```
Comment (RustyYato) [inline core/src/lib.rs:212]:
```suggestion
```
I got rid of the name iterator and forgot the update the docs here 😮

File: core/src/lib.rs [hunk 6]
```diff
@@ -310,18 +306,18 @@ pub trait ProjectAll<Parent, F> {
...
+    // `?Sized` see
+    // https://github.com/RustyYato/generic-field-projection/issues/39 for more
+    // information
 
-    /// The type that the field comes from
+    /// - type which is generating `Field`
     type Parent;
 
-    /// The type of the field itself
+    /// - a type representation of `Field` itself
```
[the remark this answers is not in the document]
Comment (RustyYato) [reply core/src/lib.rs:317]:
```suggestion
    /// a type representation of `Field` itself
```

File: derive/src/lib.rs [hunk 7]
```diff
@@ -82,7 +83,7 @@ use syn;
 /// #[allow(non_snake_case)]
 /// mod Person_fields {
 ///     use super::*;
-///     // represents the `name` field of `Person`
+///     // defines the `name` field of `PersonFields`
```
Comment (RustyYato) [inline derive/src/lib.rs:86]: This edit isn't right, please revert it. The type `name` here theoretically represents the field `name` in `Person`, and that's what I was conveying in the docs. Sure, mechanically it is a field in `PersonFields`, but that doesn't tell us *why* it exists.

File: derive/src/lib.rs [hunk 8]
```diff
@@ -106,7 +107,7 @@ use syn;
 ///             &raw mut (*ptr).name
 ///         }
 ///     }
-///     // represents the `age` field of `Person`
+///     // defines the `age` field of `PersonFields`
```
Comment (RustyYato) [inline derive/src/lib.rs:110]: Same as above

File: derive/src/lib.rs [hunk 9]
```diff
@@ -130,7 +131,7 @@ use syn;
 ///             &raw mut (*ptr).age
 ///         }
 ///     }
-///     // represents the `children` field of `Person`
+///     // defines the `children` field of `PersonFields`
```
Comment (RustyYato) [inline derive/src/lib.rs:134]: Same as above

**Review by RustyYato — CHANGES_REQUESTED** [review]
First off, thanks a lot for this massive facelift, it's great! I've noted some specific changes in this review (sorry about the massive review, but massive changes require massive reviews 😎).

A few changes that wouldn't work in the review format:
* Can you please pick between `*` or `-` for lists? Either one is fine, but not both.
* Please don't use either `*` or `-` if there is only one item in the list. I caught some of these in my edit requests, but I might have missed some.

File: core/src/lib.rs [hunk 10]
```diff
@@ -264,9 +259,9 @@ pub trait ProjectAll<Parent, F> {
 /// }
 /// ```
 ///
-/// But it is better to just derive `Field` on the types that you need to, and
-/// then use the [`chain`](trait.Fields.html#method.chain) combinator to project
-/// to the fields of fields
+/// Best practice is to derive `Field` on the needed types, using the
+/// [`chain`](trait.Fields.html#method.chain) combinator to project to the
+/// fields of `Field` rather than implementing `Field` manually
```
Comment (RustyYato) [inline core/src/lib.rs:264]: Oh, this is so much better than I was expecting!

**Review by RustyYato — APPROVED** [review]
I see you got some typos that I made, thanks a lot! This looks good to me. I'll request a review from @ckaran for a second opinion, and then this should be good to go.

File: core/src/lib.rs [hunk 11]
```diff
@@ -131,67 +131,67 @@ unsafe impl<T> Field for Identity<T> {
...
     type Projection;
 
-    /// projects to the given field
+    /// Project to a given `Field`
     fn project_to(self, field: F) -> Self::Projection;
 }
 
-// TODO: reword this documentation, tis bad
-/// Projects a type to the given field
+/// Project a given Field onto a procedurally generated Field using an `unsafe`
```
Comment (ckaran) [inline core/src/lib.rs:143]: Would it be better to quote Field?  That is, \`Field\` instead of plain 'Field'?

File: core/src/lib.rs (hunk 10 above)
Comment (ckaran) [reply core/src/lib.rs:264]: From what I've read so far, it looks great!

File: core/src/lib.rs [hunk 12]
```diff
@@ -392,12 +384,7 @@ pub unsafe trait Field {
         ptr.cast::<u8>().sub(self.field_offset()).cast()
     }
 
-    /// projects the raw pointer from the field `Type` to the `Parent` type
-    ///
-    /// # Safety
-    ///
-    /// * `ptr` must point into a valid allocation of `Type`
-    /// * `ptr` must point to a field of `Parent` with the type `Type`
+    /// Project a raw pointer from a `Type` to its generating `Parent`
```
Comment (ckaran) [inline core/src/lib.rs:387]: The safety comments here are important.  They should remain in the docs even if they are repeated elsewhere (people tend to skim docs and may miss the important bits unless they are repeated).

**Review by ckaran — CHANGES_REQUESTED** [review]
I noticed that one of the comments regarding safety was removed; it needs to be put back in. 

Other than that, it looks great!  Thank you for the PR!

File: core/src/lib.rs (hunk 12 above)
Comment (kw-fn) [reply core/src/lib.rs:387]: nice catch!

**Comment by kw-fn** [comment]
all `fmt` checks are passing locally for me

File: core/src/lib.rs [hunk 13]
```diff
@@ -408,39 +400,44 @@ pub unsafe trait Field {
...
     }
 
-    /// projects the raw pointer from the field `Type` to the `Parent` type
+    /// Project a raw pointer from a `Type` to its `Parent`
+    ///
+    /// # Safety
+    ///
+    /// * `ptr` must point into a valid allocation of `Type`
+    /// * `ptr` must point to a field of `Parent` with the type `Type`
     fn wrapping_project_raw(
```
Comment (RustyYato) [inline core/src/lib.rs:409]: This should go on `inverse_project_raw_mut`, not `wrapping_project_raw`

**Review by RustyYato — CHANGES_REQUESTED** [review]
Looks like a copy-paste error, no biggie

**Comment by RustyYato** [comment]
Looks like `rustfmt` is broken on nightly, so we can ignore that.

**Review by RustyYato — APPROVED** [review]
Ok, it looks great! Thank you @kw-fn for giving the documentation a much-needed facelift.

**Comment by RustyYato** [comment]
Whoops, accidentally clicked a button 7540ef8 wasn't meant to happen!

**Comment by kw-fn** [comment]
Thank you both for reviewing this and accepting my proposed changes, it was a
good learning experience, I'm looking forward to helping improve other parts of
the project as I can.

**Comment by RustyYato** [comment]
I'm looking forward to your next contribution. If you have any questions, you can make an issue or ask me via a message on either users.rust-lang.org or internals.rust-lang.org, with the same username.

**Pull request merged**